### PR TITLE
Twitter max length 280

### PR DESCRIPTION
**Pull request by FMRb**
Support newly change of twitter from 140 to 280 characters

https://github.com/twitter/twitter-text/issues/196

### CODE REVIEW

**Comment by saschanaz** [comment]
I think tests for invalid tweets with double-word characters also needed.

For example:

`のののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののの`

This is invalid.

**Comment by FMRb** [comment]
@saschanaz It seems to pass the test for double-word characters
`のののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののの` == 280 twitter lenght characters

**Comment by saschanaz** [comment]
Yeah, but we now need a failure test as 280 characters is not allowed for Japanese Kana characters. (CJK texts are now special. [See more](https://blog.twitter.com/official/en_us/topics/product/2017/tweetingmadeeasier.html))

File: conformance/validate.yml [hunk 1]
```diff
@@ -14,10 +14,10 @@ tests:
       expected: true
 
     - description: "Valid Tweet: 280 characters (double byte characters)"
-      text: "のののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののの"
+      text: "のののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののののの"
       expected: true
```
Comment (saschanaz) [inline conformance/validate.yml]: Should be `false`. (And this means some logic should be changed.)

**Comment by FMRb** [comment]
Thanks @saschanaz.  You are pointed out something I wasn't aware of. The library needs to change the count for Japanese, Chinese and Korean characters. It is currently counting them as single character.

**Comment by saschanaz** [comment]
Source says some rules:

image

**Comment by mariotaku** [comment]
You can learn more about 280 characters rule here: https://developer.twitter.com/en/docs/developer-utilities/twitter-text.html
I submitted a PR for Java implementation - #199. Feel free to use those code.

**Comment by saschanaz** [comment]
BTW, it seems Twitter is making twitter-text 2.0? https://twittercommunity.com/t/updating-the-character-limit-and-the-twitter-text-library/96425

**Comment by mariotaku** [comment]
@sagemintblue Ohh, that's nice. Thanks for noticing!

**Comment by lincolndbryant** [comment]
Are emojis supposed to count as 2 characters also?  No docs we found suggest this, so we filed a bug https://twittercommunity.com/t/bug-emojis-now-counting-as-2-characters-on-native-web-app/96666

We spliced in these changes for now, but they count emojis as 1.  We would love this merged ASAP to get rid of our hackish patch.

**Comment by twuttke** [comment]
I'm going to guess that the policy is all characters count double unless
they are on a list of ranges of low-density characters. Emoji are
high-density characters so they should not be counted as single, IMO.

This described at the developer link mentioned earlier in this thread

{ "version": 2, "maxWeightedTweetLength": 280, "scale": 100,
"defaultWeight": 200, "transformedURLLength": 23, "ranges": [ { "start": 0,
"end": 4351, "weight": 100 }, { "start": 8192, "end": 8205, "weight": 100
}, { "start": 8208, "end": 8223, "weight": 100 }, { "start": 8242, "end":
8247, "weight": 100 } ] }


On Nov 9, 2017 12:48 PM, "Lincoln Bryant" <notifications@github.com> wrote:

Are emojis supposed to count as 2 characters also? No docs we found suggest
this, so we filed a bug https://twittercommunity.com/
t/bug-emojis-now-counting-as-2-characters-on-native-web-app/96666

We spliced in these changes for now, but they count emojis as 1. We would
love this merged ASAP to get rid of our hackish patch.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/twitter/twitter-text/pull/198#issuecomment-343286483>,
or mute the thread

**Comment by twuttke** [comment]
that being said, the visual style modifier characters that follow some
emoji probably shouldn't count so heavily. I would even suggest a weight of
0 if that could be done in a way that doesn't introduce vulnerabilities or
bugs. I'm taking about Unicode fe0e-fe0f

**Comment by andypiper** [comment]
We'll be releasing twitter-text 2.0 in the next couple of weeks per [this announcement](https://twittercommunity.com/t/updating-the-character-limit-and-the-twitter-text-library/96425/). The logic we're using is documented here.

**Comment by hebertialmeida** [comment]
I have implemented that in Swift https://gist.github.com/hebertialmeida/d868bd58efe8b5240e4ef8b542107099

**Comment by andypiper** [comment]
twitter-text 2.0 is now available.

**Pull request closed without merging**